parla/tasks.py

```diff
@@ -447,6 +447,124 @@ def get_current_devices() -> List[Device]:
+      be supplied if any element of the list is not an array. This may be a list of \
+      devices if amount is a list of array objects.
+    """
+    from . import cpu
+    if isinstance(amount, int):
+        memsize = amount
+    elif hasattr(amount, '__cuda_array_interface__'):
+        import cupy
+        # cupy has the info we need and views are cheap,
+        # so get a view then read off the size/device info.
+        view = cupy.array(amount)
+        memsize = view.nbytes
+        if device is None:
+            device = view.device
+    else:
+        # Check if "amount" supports the buffer protocol.
+        # if it does, we're reserving memory on the CPU
+        # unless the user says otherwise. If it does not,
+        # then assume it's a list of amount parameters
+        # that need to be handled individually.
+        amount_must_be_iterable = False
+        try:
+            view = memoryview(amount)
+        except TypeError:
+            amount_must_be_iterable = True
+        else:
+            memsize = view.nbytes
+            if device is None:
+                device = cpu(0)
+        # Not a cpu array, so try handling amount as
+        # an iterable of things that each need to be processed.
+        if amount_must_be_iterable:
+            try:
+                iter(amount)
+            except TypeError as exc:
+                raise ValueError("Persistent memory spec is not an integer, array, or iterable object") from exc
+            if device is None:
+                with ExitStack() as stack:
+                    for arr in amount:
+                        inner_must_be_iterable = False
+                        try:
+                            arr.__cuda_array_interface__
+                        except AttributeError as exc:
+                            inner_must_be_iterable = True
+                        else:
+                            stack.enter_context(reserve_persistent_memory(arr))
+                        if inner_must_be_iterable:
+                            try:
+                                iter(arr)
+                            except TypeError as exc:
+                                raise ValueError("Implicit location specification only supported for GPU arrays.") from exc
+                            else:
+                                stack.enter_context(reserve_persistent_memory(arr))
+                    yield
+                    return
+            device_must_be_iterable = False
+            try:
+                device = _get_parla_device(device)
+            except ValueError:
+                device_must_be_iterable = True
+            if device_must_be_iterable:
+                with ExitStack() as stack:
+                    for arr, dev in zip(amount, device):
+                        stack.enter_context(reserve_persistent_memory(arr, dev))
+                    yield
+                    return
```

Comment:
I do NOT want to support this. It's confusing to have an implicit zip operation. If you want to have a variant of `reserve_persistent_memory` that accepts a dict or a collection of pairs that would be fine with me. Though I think it should be a different function. Too many ways to call a single function makes type errors impossible to debug. Numpy falls into this trap a lot, let's not follow them.

Reply:
IMHO the ability to expand this operation automatically over lists of things is a major value add right now since the cases where we actually use this all involve times when we're storing lists of per-device blocks. The zip seemed like the most natural way to extend the device argument to match that use-case. We can get rid of the zip, but recursively expanding over lists doesn't fit in the list of functions you've provided below unless we generalize reserve_data a bit.

Reply:
I'm OK with recursively flattening the arguments to `reserve_data`. The zip is the confusing part to me. I would be a little happier with cartesian product and relatively comfortable with implicitly mapping over *one* argument as long as the other is a single value.

Reply:
Yah, getting rid of the zip option is totally fine with me.